File: 3_Examples/multiControl.cpp
#include <iostream>
#include <cudaq.h>
#include <cudaq/algorithm.h>
#include <cudaq/algorithms/draw.h>
#include <cudaq/spin_op.h>

__qpu__ void x_kernel(cudaq::qubit &qubit)
{
    x(qubit);
}

__qpu__ void kernel()
{
    auto control_vector = cudaq::qvector(2);
    auto target = cudaq::qubit();

    cudaq::x(control_vector);
    cudaq::x(target);
    cudaq::x(control_vector[1]);
    cudaq::control(x_kernel, control_vector, target);
}

int main(int argc, char const *argv[])
{
    std::cout << cudaq::draw(kernel);

    auto results = cudaq::sample(1000, kernel);
    results.dump();

    return 0;
}

File: 3_Examples/multiControl.py
import cudaq
import numpy as np


# A kernel that performs an X-gate on a provided qubit.
@cudaq.kernel
def x_kernel(qubit: cudaq.qubit):
    x(qubit)


# A kernel that will call `x_kernel` as a controlled operation.
@cudaq.kernel
def kernel():
    control_vector = cudaq.qvector(2)
    target = cudaq.qubit()
    x(control_vector)
    x(target)
    x(control_vector[1])
    cudaq.control(x_kernel, control_vector, target)


print(cudaq.draw(kernel))
results = cudaq.sample(kernel)
print(results)


@cudaq.kernel
def kernel():
    qvector = cudaq.qvector(3)
    x(qvector)
    x(qvector[1])
    x.ctrl([qvector[0], qvector[1]], qvector[2])
    mz(qvector)


print(cudaq.draw(kernel))
results = cudaq.sample(kernel)
print(results)
